File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/templateengine/blocks/Block.java
// Copyright (c) Microsoft. All rights reserved.
package com.microsoft.semantickernel.templateengine.blocks;

/** Base class for blocks parsed from a prompt template */
public abstract class Block {
    private final String content;
    private final BlockTypes type;

    /**
     * Base constructor
     *
     * @param content Block content
     * @param type Block type
     */
    public Block(String content, BlockTypes type) {
        if (content == null) {
            content = "";
        }

        this.content = content;
        this.type = type;
    }

    /**
     * Get the block content
     *
     * @return Block content
     */
    public String getContent() {
        return content;
    }

    /**
     * Check if the block content is valid.
     *
     * @return True if the block content is valid
     */
    public abstract boolean isValid();

    /**
     * Get the block type
     *
     * @return Block type
     */
    public BlockTypes getType() {
        return type;
    }
}

File: dotnet/src/Extensions/TemplateEngine.PromptTemplateEngine/Blocks/Block.cs
// Copyright (c) Microsoft. All rights reserved.

using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace Microsoft.SemanticKernel.TemplateEngine.Prompt.Blocks;

/// <summary>
/// Base class for blocks parsed from a prompt template
/// </summary>
public abstract class Block
{
    internal virtual BlockTypes Type => BlockTypes.Undefined;

    // internal virtual bool? SynchronousRendering => null;

    /// <summary>
    /// The block content
    /// </summary>
    internal string Content { get; }

    /// <summary>
    /// App logger
    /// </summary>
    private protected ILogger Logger { get; }

    /// <summary>
    /// Base constructor. Prevent external instantiation.
    /// </summary>
    /// <param name="content">Block content</param>
    /// <param name="loggerFactory">The <see cref="ILoggerFactory"/> to use for logging. If null, no logging will be performed.</param>
    private protected Block(string? content, ILoggerFactory? loggerFactory)
    {
        this.Content = content ?? string.Empty;
        this.Logger = loggerFactory is not null ? loggerFactory.CreateLogger(this.GetType()) : NullLogger.Instance;
    }

    /// <summary>
    /// Check if the block content is valid.
    /// </summary>
    /// <param name="errorMsg">Error message in case the content is not valid</param>
    /// <returns>True if the block content is valid</returns>
    public abstract bool IsValid(out string errorMsg);
}
